#1103: Check how authentication attributes are restored... [master branch]

jkacer opened the pull request:
This is a pull request to solve issue #1103 for the master branch - port of changes from branch 2.2.x.
I have enhanced the ProfileHelper and the ProfileDefinition with a few new methods that support authentication attributes.
Existing methods, deprecated in 2.2.x and 2.3.x, were removed from 3.0.x. Calls from clients and unit tests were adapted.

jkacer commented:
I'm not sure if I should care about the warnings from Code Climate, please tell me what you think. Thanks.

leleuj commented on pac4j-cas/src/main/java/org/pac4j/cas/credentials/authenticator/CasAuthenticator.java:
```diff
@@ -90,9 +90,9 @@ public void validate(final TokenCredentials credentials, final WebContext contex
                 profile = getProfileDefinition().newProfile(principal, configuration.getProxyReceptor());
                 profile.setId(ProfileHelper.sanitizeIdentifier(profile, id));
-                getProfileDefinition().convertAndAdd(profile, newAttributes);
+                getProfileDefinition().convertAndAdd(profile, newAttributes, null);
```
Instead of `null`, we should use the authentication attributes retrieved via `assertion.getAttributes()`.

jkacer replied:
Aha, OK, I'll have a look. I generally adapted all the calls by using profile attributes only to make it compile and act exactly like now (without authentication attributes). But you are right, we can improve it by using the authentication attributes where possible, since they are already available.

leleuj commented:
You can generally ignore the warnings from code climate, or at least that's what I do most of the time ;-)

One minor comment and a question and we'll be good.

jkacer commented:
Hi! I have modified `CasAuthenticator` to count with authentication attributes as you suggested. I was not sure if they need the undergo the same transformation by the attribute handler but I decided to pass them through it too.
Concerning other possible uses, I was not able to find anything else, including SAML. In `SAML2Authenticator.validate()`, we operate with the attributes a lot but all attributes from `SAML2Credentials` are implicitly considered as profile attributes; authentication attributes are explicitly enumerated and added one-by-one. But if you are able to find a concrete call that could be updated, please tell me.
Would you like to have the fix of CAS Authenticator in 2.2.x too?

leleuj commented:
OK. Thanks for checking.

Regarding the 2.2.x branch, do it if you find some time, but it's not a priority.

leleuj commented:
Thank you for contributing!
